### PR TITLE
configDB enforcer for interface

### PR DESCRIPTION
**Pull request by JipanYanga**
Support VLAN interface only for now

Signed-off-by: Jipan Yang <jipan.yang@alibaba-inc.com>

What I did
Configuration enforcer for VLAN interface

Why I did it
Support interface configuration via configDB for both initialization and incremental update
How I verified it
VLAN trunk test plan: https://github.com/Azure/SONiC/wiki/VLAN-trunk-test-plan
Details if related

### CODE REVIEW

**Comment by JipanYanga** [comment]
Has dependency on:
Azure/sonic-swss-common#146
#355
#336
#344

**Comment by lguohan** [comment]
@taoyl-ms , @stcheng  to review

File: intfconfd/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+bin_PROGRAMS = intfconfd
+
+if DEBUG
+DBGFLAGS = -ggdb -DDEBUG
+else
+DBGFLAGS = -g
+endif
+
+intfconfd_SOURCES = intfconfd.cpp intfconf.cpp $(top_srcdir)/cfgorch/cfgorch.cpp
```
Comment (qiluo-msft) [inline intfconfd/Makefile.am]: >cfgorch [](start = 61, length = 7)

suggest merge "cfgorch", "intfconfd" into one directory

File: intfconfd/intfconf.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+#include "intfconf.h"
+#include "exec.h"
+
+using namespace std;
+using namespace swss;
+
+#define VLAN_PREFIX         "Vlan"
+#define LAG_PREFIX          "PortChannel"
+
+IntfConf::IntfConf(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, vector<string> tableNames) :
```
Comment (qiluo-msft) [inline intfconfd/intfconf.cpp]: >vector [](start = 81, length = 6)

Do you want to copy a vector?

File: intfconfd/intfconf.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    CfgOrch::syncCfgDB(CFG_VLAN_INTF_TABLE_NAME, m_cfgVlanIntfTable);
+}
+
+bool IntfConf::setIntfIp(string &alias, string &opCmd, string &ipPrefixStr)
+{
+    string cmd, res;
+
+    cmd = "ip address " + opCmd + " ";
+    cmd += ipPrefixStr + " dev " + alias;
+    swss::exec(cmd, res);
```
Comment (qiluo-msft) [inline intfconfd/intfconf.cpp]: >exec [](start = 10, length = 4)

check the return value?

File: intfconfd/intfconf.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+}
+
+void IntfConf::syncCfgDB()
+{
+    CfgOrch::syncCfgDB(CFG_INTF_TABLE_NAME, m_cfgIntfTable);
+    CfgOrch::syncCfgDB(CFG_VLAN_INTF_TABLE_NAME, m_cfgVlanIntfTable);
+}
+
+bool IntfConf::setIntfIp(string &alias, string &opCmd, string &ipPrefixStr)
```
Comment (qiluo-msft) [inline intfconfd/intfconf.cpp]: >string & [](start = 25, length = 8)

If using a reference, consider using a const reference if possible.

File: intfconfd/intfconf.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    cmd += ipPrefixStr + " dev " + alias;
+    swss::exec(cmd, res);
+    return true;
+}
+
+bool IntfConf::isIntfStateOk(string &alias)
+{
+    vector<FieldValueTuple> temp;
+
+    if (!alias.compare(0, strlen(VLAN_PREFIX), VLAN_PREFIX))
```
Comment (qiluo-msft) [inline intfconfd/intfconf.cpp]: >alias [](start = 9, length = 5)

alias == VLAN_PREFIX

File: intfconfd/intfconf.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+{
+    string cmd, res;
+
+    cmd = "ip address " + opCmd + " ";
+    cmd += ipPrefixStr + " dev " + alias;
+    swss::exec(cmd, res);
+    return true;
+}
+
+bool IntfConf::isIntfStateOk(string &alias)
```
Comment (qiluo-msft) [inline intfconfd/intfconf.cpp]: >isIntfStateOk [](start = 15, length = 13)

suggest name: hasIntf

File: intfconfd/intfconf.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        string alias(keys[0]);
+
+        if (alias.compare(0, strlen(VLAN_PREFIX), VLAN_PREFIX))
+        {
+            /* handle IP over vlan Only for now, skip the rest */
+            it = consumer.m_toSync.erase(it);
+            continue;
+        }
+
+        IpPrefix ip_prefix(kfvKey(t).substr(kfvKey(t).find(CONFIGDB_KEY_SEPARATOR)+1));
```
Comment (qiluo-msft) [inline intfconfd/intfconf.cpp]: >IpPrefix ip_prefix(kfvKey(t).substr(kfvKey(t).find(CONFIGDB_KEY_SEPARATOR)+1)); [](start = 8, length = 79)

Split into multiple statement. It is too complex to read.

File: intfconfd/intfconfd.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                if (o->hasSelectable(c))
+                {
+                    o->execute(c->getTableName());
+                }
+            }
+        }
+    }
+    catch(const std::exception &e)
+    {
+        SWSS_LOG_ERROR("Runtime error: %s", e.what());
```
Comment (qiluo-msft) [inline intfconfd/intfconfd.cpp]: return a non zero error code?

File: intfconfd/Makefile.am (hunk 1 above)
Comment (JipanYanga) [reply intfconfd/Makefile.am]: Yes,  cfgagent might be a good name for the directory.  Will move intfconfd, vlanconfd and debug cli implmentation cfgmgrtest there.  Actually that was the initial arrangement.

File: intfconfd/intfconf.cpp (hunk 2 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: The table names for different DB are different. a vector might not be a good fit here.

File: .gitignore [hunk 9]
```diff
@@ -43,6 +43,7 @@ deps/
 teamsyncd/teamsyncd
 fpmsyncd/fpmsyncd
 intfsyncd/intfsyncd
+cfgagent/intfconfd
```
Comment (lguohan) [inline .gitignore]: can you rename to intfmgr? 

The role of confd, orchagent, syncd are blurred, it is hard to have a clear cuts of different roles they are playing, so mgr can include all functions, which manages both asic and linux setup. therefore I prefer mgr name

File: cfgagent/intfconf.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+#include "intfconf.h"
+#include "exec.h"
+
+using namespace std;
+using namespace swss;
+
+#define VLAN_PREFIX         "Vlan"
+#define LAG_PREFIX          "PortChannel"
+
+IntfConf::IntfConf(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, vector<string> tableNames) :
```
Comment (lguohan) [inline cfgagent/intfconf.cpp]: as requested above, can you rename to IntfMgr

**Comment by lguohan** [comment]
can you add swssloglevel support for this agent so that we can control the log level of this agent on the fly.

File: intfconfd/intfconf.cpp (hunk 2 above)
Comment (qiluo-msft) [reply intfconfd/intfconf.cpp]: I am not sure you are talking about the same thing. I mean that use a copied instance may cost much than a reference.

File: cfgagent/Makefile.am [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+bin_PROGRAMS = intfconfd
+
+if DEBUG
+DBGFLAGS = -ggdb -DDEBUG
+else
+DBGFLAGS = -g
+endif
+
+
+intfconfd_SOURCES = intfconfd.cpp intfconf.cpp $(top_srcdir)/orchagent/orchbase.cpp
```
Comment (qiluo-msft) [inline cfgagent/Makefile.am]: >orchbase [](start = 71, length = 8)

Is this file missing in the PR? Why not added it in the same directory?

**Comment by lguohan** [comment]
retest this please

File: cfgagent/intfconf.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+bool IntfConf::isIntfStateOk(string &alias)
+{
+    vector<FieldValueTuple> temp;
+
+    if (!alias.compare(0, strlen(VLAN_PREFIX), VLAN_PREFIX))
+    {
+        if (m_stateVlanTable.get(alias, temp))
+        {
+            SWSS_LOG_DEBUG("Port %s is ready\n", alias.c_str());
```
Comment (lguohan) [inline cfgagent/intfconf.cpp]: Vlan is ready.

File: cfgagent/intfconf.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+            continue;
+        }
+        IpPrefix ip_prefix(kfvKey(t).substr(pos+1));
+
+        SWSS_LOG_DEBUG("intfs doTask: %s", (dumpTuple(consumer, t)).c_str());
+
+        string op = kfvOp(t);
+        if (op == SET_COMMAND)
+        {
+            /* Don't proceed if port/lag/VLAN is not ready yet */
```
Comment (lguohan) [inline cfgagent/intfconf.cpp]: add more comments here, retry. Later, we need to add subscriber for the state db to get notification when the port/lag/vlan is created.

File: .gitignore (hunk 9 above)
Comment (JipanYanga) [reply .gitignore]: Ok, will rename them to intfmgr and vlanmgr.  For switch table level configuration,  besides switchmgr process itself,  the configuration may be used by multiple function modules like VLAN.

I plan to keep the class name of VlanConf, IntfConf and SwitchConf,  let me know if you have other suggestions.

File: cfgagent/intfconf.cpp (hunk 10 above)
Comment (JipanYanga) [reply cfgagent/intfconf.cpp]: So process and binary name will be changed to VlanMgr/IntfMgr/SwitchMgr,  you want to change the internal class name to xxxMgr too?   IntfConfd/VlanConfd were chosen per the offline meeting discussion.  Yes, class name may be changed.

File: cfgagent/Makefile.am (hunk 11 above)
Comment (JipanYanga) [reply cfgagent/Makefile.am]: It is in PR   https://github.com/Azure/sonic-swss/pull/360

File: cfgagent/intfconf.cpp (hunk 12 above)
Comment (JipanYanga) [reply cfgagent/intfconf.cpp]: ok, will fix it.

File: cfgagent/intfconf.cpp (hunk 13 above)
Comment (JipanYanga) [reply cfgagent/intfconf.cpp]: ok

**Comment by JipanYanga** [comment]
@lguohan lguohan   swssloglevel has supported with "Logger::linkToDbNative("intfconfd");"

root@sonic:/etc/sonic# swssloglevel -p
SAI_API_ACL                   SAI_LOG_LEVEL_NOTICE
........
**intfconfd                     NOTICE**
intfsyncd                     NOTICE
..........
**vlanconfd                     DEBUG**

File: intfconfd/intfconf.cpp (hunk 2 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: ok

File: intfconfd/intfconf.cpp (hunk 4 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: ok

File: intfconfd/intfconf.cpp (hunk 3 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: Error has been logged in exec if there is any. In this case, error is not really expected.
That is common question for many functions in sonic, end 2 end feedback loop is missing.

Comment (qiluo-msft) [reply intfconfd/intfconf.cpp]: I don't agree. Locally if there is something error happens, you should check the return code all the levels up the stack, and quit as soon as possible.

File: intfconfd/intfconf.cpp (hunk 6 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: Honestly I don't see readyIntf a better name than isIntfStateOk,  we do put state ok in stateDB though it is checked  for now.  If you insist on changing the name, I could do it.

Comment (qiluo-msft) [reply intfconfd/intfconf.cpp]: Just checking the code, there is nothing about "Ok" state. So I guess the original name is not perfect.

File: intfconfd/intfconf.cpp (hunk 3 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: fine.  But I don't see the same criteria applied in orchagent, syncd, those python code and etc.
We could start the endeavor from this.

File: intfconfd/intfconf.cpp (hunk 6 above)
Comment (JipanYanga) [reply intfconfd/intfconf.cpp]: Ok,  will add code to check the state:

127.0.0.1:6379[6]> hgetall "PORT_TABLE|Ethernet47"
1) "state"
2) "ok"

File: cfgagent/intfmgr.cpp [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        m_cfgIntfTable(cfgDb, CFG_INTF_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_cfgVlanIntfTable(cfgDb, CFG_VLAN_INTF_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_statePortTable(stateDb, STATE_PORT_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_stateLagTable(stateDb, STATE_LAG_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_stateVlanTable(stateDb, STATE_VLAN_TABLE_NAME, CONFIGDB_TABLE_NAME_SEPARATOR),
+        m_appIntfTableProducer(appDb, APP_INTF_TABLE_NAME)
+{
+}
+
+void IntfMgr::syncCfgDB()
```
Comment (lguohan) [inline cfgagent/intfmgr.cpp]: where does this function get called?

File: configure.ac [hunk 15]
```diff
@@ -87,6 +87,7 @@ AC_CONFIG_FILES([
     teamsyncd/Makefile
     swssconfig/Makefile
     tests/Makefile
+    cfgagent/Makefile
```
Comment (lguohan) [inline configure.ac]: >cfgagent [](start = 4, length = 8)

cfgmgr

**Comment by JipanYanga** [comment]
Switched to orch class.
Has dependency on https://github.com/Azure/sonic-swss/pull/356

File: cfgmgr/intfmgr.cpp [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    stringstream cmd;
+    string res;
+
+    cmd << IP_CMD << " address " << opCmd << " " << ipPrefixStr << " dev " << alias;;
+    int ret = swss::exec(cmd.str(), res);
+    if (ret == 0)
+    {
+        return true;
+    }
+    return false;
```
Comment (qiluo-msft) [inline cfgmgr/intfmgr.cpp]: return ret == 0;

File: intfconfd/intfconf.cpp (hunk 6 above)
Comment (qiluo-msft) [reply intfconfd/intfconf.cpp]: I don't see that you check "ok"

**Comment by lguohan** [comment]
@JipanYanga , can you resolve the conflict?

File: cfgmgr/intfmgr.cpp (hunk 16 above)
Comment (JipanYanga) [reply cfgmgr/intfmgr.cpp]: done.

**Comment by JipanYanga** [comment]
Conflicts resolved.

**Pull request merged**